Up to 2x faster _representative_index_to_coefficient

ll-ara opened the pull request:
I have had a crack a optimising the _representative_index_to_coefficient function. Note that this might partially solve issue #1216. 

The only time that I could find that it was called was in the unit tests so I have based my optimisation on that. For “run_spineopt_representative_periods.jl”, _representative_index_to_coefficient took around 3 seconds on my machine to execute, it now takes 0.7 seconds. Hopefully this improvement scales to the EU case study which I don’t have access to yet (please check on my behalf).

I don’t think any extra extra unit tests are required since the only changes are internal to the function. If some unit tests are required then I would love some help writing them.


## Checklist before merging
- [x] Documentation is up-to-date
- [x] Code has been formatted according to SpineOpt's style
- [x] Unit tests pass
- [x] Verified to improve performance of EU case study mentioned in #1216 (please run on my behalf)
- [ ]  Write unit test
- [ ] Fix bug in code

datejada commented:
@ll-ara Thanks for the PR, some comments:

- Please rebase to the latest changes in master so that we can get the tests running for this PR.
- I would like to understand better how you got the speed up. Would you mind explaining more about the why of your changes? A brief description is enough, for instance, type stability.
- I attached the data to the issue. Can you comment on the speed you get in that case?

ll-ara commented on src/variables/variable_common.jl:
```diff
@@ -285,14 +285,35 @@ A `Dict` mapping representative indidces to coefficient.
-        !any(isempty.(keys(representative_inds_to_coef)))
+    representative_inds_to_coef = nothing
+    # variables to cache the values so that the types can be known ahead of time
+    indCache = nothing
```
yes of course, sorry will change that

ll-ara commented:
The code I originally wrote to speed up the unit test didn’t speed up the provided case study. This isn’t really that surprising given that it only takes two orders of magnitude go from L1 cache to L3 and the increase in input size was around three orders of magnitude larger (input file size when from ~20kB to ~20MB). 

I’m using the same ideas as before to reduce the amount of computation by effectively turning three loops into one loop (+ a trivial one). However this doesn’t necessarily help because memory is the new bottleneck. I suspect that type stability is a large contributing factor here and this is an area that I haven’t really delved into but we definitely should. For example, Julia doesn’t seem to know what type representative_t_to_coef_arr immediately is at runtime. I think we should do what JUMP does and explicitly type every input to all functions. 

The large improvement that I’m using here is by using @memoize to cache values. Along with the other optimisations this decreased the time it takes to run _representative_index_to_coefficient from ~175 seconds to ~80 seconds (~2x). Note that of this 80 seconds, around 65 seconds (~80% of the time) is just spent in get_ind_cached(). This demonstrates that it isn’t (or at least no longer) _representative_index_to_coefficient that is the issue and rather child functions and type stability. 

Using @memoize is definitely a cheat way of improving performance. I’m not solving the many of the underlying issues but rather restructuring the code to avoid them. I’m not really suggesting this should get merged – unless the improvement is a net benefit in the short term.

ll-ara commented:
It seems that I refactored the changes in the wrong order so that it looks like I changed more files than I actually did. I assume I can fix this by re-writing history and force pushing? 

I'm assuming that the failing CI for Julia 1.8 on windows is somehow related to #1222 but I could be wrong. 

I'm a bit confused why the codecov/project is failing? Does the "1 missed lines" mean that it was covered before but now is no longer covered?

ll-ara commented:
Unfortunately I made a mistake (naming conflict) in my code and adding `@memoize` actually didn’t make it faster at all. Instead it was skipping a whole iteration of the loop which wasn’t picked up by the unit tests. Explicit input types would have also helped in this particular case.

I did end up writing a unit test case for a few lines of code that wasn’t covered (as discussed above) so will submit that in a separate pull request to avoid confusion. 

I tried a few other things to improve performance, including adding explicitly adding input types (which seemed to help a little) but ultimately I didn’t find anything really noteworthy. Thanks for the help everyone, closing this draft issue.
